libraries/AP_RangeFinder/AP_RangeFinder_Benewake.cpp

```diff
@@ -0,0 +1,165 @@
+// byte 6 (TF02)        SIG             Reliability in 8 levels, 7 & 8 means reliable
+// byte 6 (TFmini)      Distance Mode   0x02 for short distance (mm), 0x07 for long distance (cm)
+// byte 7 (TF02 only)   TIME            Exposure time in two levels 0x03 and 0x06
+// byte 8               Check           Checksum parity bit, sum of byte 0 to byte 7
+
+/* 
+   The constructor also initialises the rangefinder. Note that this
+   constructor is not called until detect() returns true, so we
+   already know that we should setup the rangefinder
+*/
+AP_RangeFinder_Benewake::AP_RangeFinder_Benewake(RangeFinder::RangeFinder_State &_state,
+                                                             AP_SerialManager &serial_manager,
+                                                             uint8_t serial_instance,
+                                                             benewake_model_type model) :
+    AP_RangeFinder_Backend(_state),
+    model_type(model)
+{
+    uart = serial_manager.find_serial(AP_SerialManager::SerialProtocol_Rangefinder, serial_instance);
+    if (uart != nullptr) {
+        uart->begin(serial_manager.find_baudrate(AP_SerialManager::SerialProtocol_Rangefinder, serial_instance));
+    }
+}
+
+/* 
+   detect if a Benewake rangefinder is connected. We'll detect by
+   trying to take a reading on Serial. If we get a result the sensor is
+   there.
+*/
+bool AP_RangeFinder_Benewake::detect(AP_SerialManager &serial_manager, uint8_t serial_instance)
+{
+    return serial_manager.find_serial(AP_SerialManager::SerialProtocol_Rangefinder, serial_instance) != nullptr;
+}
+
+// distance returned in reading_cm, signal_ok is set to true if sensor reports a strong signal
+bool AP_RangeFinder_Benewake::get_reading(uint16_t &reading_cm, bool &signal_ok)
+{
+    if (uart == nullptr) {
+        return false;
+    }
+
+    float sum_cm = 0;
+    uint16_t count = 0;
+    bool dist_reliable = false;
+
+    // read any available lines from the lidar
+    int16_t nbytes = uart->available();
+    while (nbytes-- > 0) {
+        char c = uart->read();
+        // if buffer is empty and this byte is 0x59, add to buffer
+        if (linebuf_len == 0) {
+            if (c == BENEWAKE_FRAME_HEADER) {
+                linebuf[linebuf_len++] = c;
+            }
+        } else if (linebuf_len == 1) {
+            // if buffer has 1 element and this byte is 0x59, add it to buffer
+            // if not clear the buffer
+            if (c == BENEWAKE_FRAME_HEADER) {
+                linebuf[linebuf_len++] = c;
+            } else {
+                linebuf_len = 0;
+                dist_reliable = false;
+            }
+        } else {
+            // add character to buffer
+            linebuf[linebuf_len++] = c;
+            // if buffer now has 9 items try to decode it
+            if (linebuf_len == BENEWAKE_FRAME_LENGTH) {
+                // calculate checksum
+                uint16_t checksum = 0;
```

Comment:
On the design sheet the checksum is written as the lower 8 bits. I think that it is good to add 8 bits and 1 byte.